Task 2

Danila-Peshev opened the pull request:
task-2

Danila-Peshev committed:
Initial commit

Danila-Peshev committed:
form login si draw (no logic)

Danila-Peshev committed:
😁

Danila-Peshev committed:
😁

Danila-Peshev committed:
😁

Danila-Peshev committed:
create login form, basic authorization has been created from the task

Danila-Peshev committed:
delete image

Isidora25 commented on src/fillLocalStorage.tsx:
```diff
@@ -0,0 +1,26 @@
+const Users = [
```
```suggestion
const USERS = [
```
в PascalCase именуются либо классы, либо компоненты

Isidora25 commented on src/index.tsx:
```diff
@@ -1,9 +1,49 @@
+
+const store = new Store();
+
+export const Context = createContext<State>({
+  store,
+})
+
+
+const router = createBrowserRouter([
+  {
+    path: '/',
+    element: (
+      <ProtectedRoute>
+        <Home />
+      </ProtectedRoute>
+    ),
+  },
+  {
+    path: '/login',
+    element: <Login />,
+  },
+]);
+
```
вынеси в отдельный файл - это улучшит читаемость

Isidora25 commented on src/index.tsx:
```diff
@@ -1,9 +1,49 @@
+
+const store = new Store();
+
+export const Context = createContext<State>({
```
подскажи, а для чего тебе нужен store? 
ты легко реализуешь логику работы и проверки текущего юзера с помощью только контекста в реакте 
кроме того, потом когда ты будешь прикручивать бэк, ты будешь использовать apollo-client  с хуками и там тоже можно прекрасно обойтись без сторов

Danila-Peshev replied:
В интернете примерах увидел много пример со стором. Сейчас всё на контексте.

Isidora25 commented on src/pages/Home/Home.tsx:
```diff
@@ -0,0 +1,26 @@
+  return (
+      <div className='w-full text-center mt-96'>
+        <div>
+        <h1 className='text-4xl font-bold'>Пользователь {store.user.email}, добро пожаловать на сайт "Моё любимое кино"!</h1>
```
добавь интернационализацию в проект с помощью lingui или i18n

Isidora25 commented on src/pages/Home/Home.tsx:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+      <div className='w-full text-center mt-96'>
+        <div>
```
этот div кажется лишний здесь

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  return (
+    <div className="w-1/4 mt-20 mx-auto">
+      <img src={logo} className="mx-auto size-40 mb-10 rounded-full"></img>
```
```suggestion
      <img src={logo} className="mx-auto size-40 mb-10 rounded-full" />
```
плюс нужно добавлять alt аттрибут  к изображениям

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,70 @@
+      <img src={logo} className="mx-auto size-40 mb-10 rounded-full"></img>
+      <div className="h-80 bg-white border border-zinc-500 shadow-xl rounded">
+        <div className="my-10 w-11/12 mx-auto">
+          <div className="flex flex-col mx-auto">
+            <label className="left">Имя пользователя или e-mail</label>
+            <input
+              value={email}
+              onChange={(e) => { setEmail(e.target.value) }}
+              type="text"
+              className="border border-neutral-300 shadow-sm h-10 text-lg"></input>
+          </div>
+          <div className="flex flex-col mx-auto my-5">
+            <label>Пароль</label>
+            <input
+              value={password}
+              onChange={(e) => { setPassword(e.target.value) }}
+              type="password"
+              className="border border-neutral-300 shadow-sm h-10 text-2xl"></input>
```
вот тут ты придумал велосипед немного 
это самая обычная форма 
для работы с формами в реакт удобно использовать react-final-form например

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,70 @@
+            <p className='text-red-500 font-semibold'>{messageInputFields}</p>
+          </div>
+          <div className="flex relative mb-10 items-center">
+            <input
+              type="checkbox"
+              checked={checked}
+              onChange={() => setChecked(!checked)}
+              className="size-5 rounded-none border-none border-neutral-300 mr-4">
+            </input>
+            <label>Запомнить меня</label>
```
это не надо реализовывать

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,70 @@
+              className="size-5 rounded-none border-none border-neutral-300 mr-4">
+            </input>
+            <label>Запомнить меня</label>
+            <div className="grow"></div>
```
это точно здесь нужно?

Danila-Peshev replied:
Он служил как распорка, чтобы кнопка была в правой стороне. Т.к. "Запомнить меня" сейчас нет, то его тоже убрал.

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -0,0 +1,70 @@
+          </div>
+        </div>
+      </div>
+      <div className="mt-5">
+        <a href="#" className="ml-5">Забыли пароль?</a>
+      </div>
```
это тоже пропускаем

Isidora25 commented on src/services/AuthService.ts:
```diff
@@ -0,0 +1,21 @@
+
+export default class AuthService {
+    static login(email: string, password: string) {
+        if (!email || !password) {
+            return { error: "Ни одно поле не должно быть пустым" };
+        }
```
это работа валидации в форме, а не сервиса аутентификации

Isidora25 commented on src/services/AuthService.ts:
```diff
@@ -0,0 +1,21 @@
+import { findByEmail } from '../repositories/UserRepository'
+
+export default class AuthService {
```
так как ты используешь функциональные компоненты, не стоит миксовать стили и использовать классы 
тут отлично подойдет просто функция или хук

Danila-Peshev replied:
Прошу прощения, после java тяжело привыкнуть к этому😅

Isidora25 commented on src/services/AuthService.ts:
```diff
@@ -0,0 +1,21 @@
+        }
+
+        const user = findByEmail(email);
+        if (!user) {
+            return { error: "Пользователя с таким email не существует" };
+        }
```
а это очень небезопасная штука. когда ты показываешь пользователю, что у тебя нет в базе такой почты, ты даешь возможность брут форсом подобрать имейлы из твоей базы 
так что лучше возвращать обобщенное сообщение об ошибке

Danila-Peshev replied:
Я старался всё делать по ТЗ, я думал это очень важно, поэтому и сделал как во вложенном примере.

Danila-Peshev committed:
corrected via comments

Isidora25 commented on public/locales/en/translation.json:
```diff
@@ -0,0 +1,9 @@
+    "login": "Login",
+    "input-email": "User name or e-mail",
+    "input-password": "Password",
+    "input-error-empty": "No field should be empty",
```
это не критично, но обычно ключи в camelCase в json

Isidora25 commented on src/components/AuthContext.tsx:
```diff
@@ -0,0 +1,50 @@
+    logout: () => void;
+}
+
+const AuthContext = createContext<AuthContextType>({} as AuthContextType);
```
`as` это всегда такой хак, в котором нужно быть 100 процентов уверенным
если ты уверен, что пустой объект не выстрелит тебе в ногу в моменты обращения к нему за переменной - это хорошо 
но лучше не злоупотреблять этим хаком и инициализировать контекст с falsy значениями, а обработчики сделать необязательными

Isidora25 commented on src/components/switch_language/LanguageSwitcher.tsx:
```diff
@@ -0,0 +1,35 @@
+    const [activeLanguage, setActiveLanguage] = useState('ru');
+    const { i18n } = useTranslation();
+  
+    const changeLanguage = useCallback((lng: string) => {
```
если changeLanguage не используется нигде кроме useEffect не выноси это из него, не создавай доп зависимости для перерендера

Isidora25 commented on src/components/switch_language/LanguageSwitcher.tsx:
```diff
@@ -0,0 +1,35 @@
+import SwitchLanguageButton from "./SwitchLanguageButton";
+
+const LanguageSwitcher = () => {
+    const [activeLanguage, setActiveLanguage] = useState('ru');
```
тут классно подойдет использование union type из typeScript 
```suggestion
type Language = 'ru' | 'en'


    const [activeLanguage, setActiveLanguage] = useState<Language>('ru');
```

тогда в activeLanguage не попадет ничего кроме 'ru' или 'en'

Isidora25 commented on src/components/switch_language/SwitchLanguageButton.tsx:
```diff
@@ -0,0 +1,26 @@
+import { FC } from "react";
+
+interface SwitchLanguageButtonProps {
+    language: string;
+    active: boolean;
+    onClick: () => void;
+  }
+  
+  interface SwitchLanguageButtonProps {
+    language: string;
+    active: boolean;
+    onClick: () => void;
+  }
```
два раза одно и то же

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -1,70 +1,92 @@
-    const result = store.login(email, password);
-    if (result.error) {
-      setMessageInputFields(result.error);
+  const onSubmit = async (values: { email: any; password: any; }) => {
```
`any` нужно использовать так же аккуратно как `as`
в данном случае ты просто выключил тайп скрипт 
только в крайних случаях, когда ты не знаешь, например, что приходит из апи 
а в этом случае ты точно знаешь, как будет тип у email  и password

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -1,70 +1,92 @@
-      setMessageInputFields(result.error);
+  const onSubmit = async (values: { email: any; password: any; }) => {
+    const { email, password } = values;
+    if (!email || !password) {
+      return { email: t("input-error-empty"), password: t("input-error-empty") };
+    }
```
лучше такого плана валидацию делать с помощью функции на филд validate  и блокировать кнопку submit пока форма не будет валидной полностью 
иначе это не оч юзер френдли 
юзер понимает, что данные не валидные только после нажатия на кнопку сабмит

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -1,70 +1,92 @@
+                  </Field>
+                </div>
+                <div className="flex flex-col mx-auto my-5">
+                  <label>{t("input-password")}</label>
+                  <Field name="password">
+                    {({ input, meta }) => (
+                      <div>
+                        <input
+                          {...input}
+                          type="password"
+                          className="border border-blue-200 shadow-sm h-10 text-2xl w-full"
+                        />
+                        {meta.touched && meta.error && (
+                          <span className="text-red-500 font-semibold">{meta.error}</span>
+                        )}
+                      </div>
+                    )}
+                  </Field>
```
DRY 
Сделай компонент, в котором будет label  и input + сообщение об ошибке и передавай в prop component в Field

Isidora25 commented on src/components/switch_language/SwitchLanguageButton.tsx:
папки лучше называть в kabab-case

Danila-Peshev committed:
Everything except the form

Danila-Peshev committed:
corrected via comments

Danila-Peshev committed:
corrected via comments

Isidora25 commented on src/pages/Login/Login.tsx:
```diff
@@ -4,87 +4,66 @@ import { useLocation, useNavigate } from 'react-router-dom';
                   <button
                     type="submit"
-                    className="bg-blue-800 h-10 w-20 rounded text-white hover:bg-white hover:text-black hover:border hover:border-black">
+                    disabled={submitting || pristine}
```
кажется, все кнопка все еще активна, если форма невалидна

Danila-Peshev committed:
corrected via comments

Danila-Peshev committed:
corrected via comments

Danila-Peshev committed:
corrected via comments

reshetov commented on package.json:
```diff
@@ -10,8 +10,17 @@
     "@types/node": "^16.18.101",
```
Стразу - коммит как и ветка должны называться понятно, можно длинно, но обязательно - понято
`task5`  ровно как и `small-fixes`, `my-last-commit-in-this-week` не несут в себе смыслвой нагрузки и по ним нельзя понять что именно было сделано. А это ОЧЕНЬ может пригодится если тебе нужно найти "кто сломал эту кнопку полгода назад"

reshetov commented on src/components/AuthContext.tsx:
```diff
@@ -0,0 +1,57 @@
+const AuthContext = createContext<AuthContextType>(defaultAuthContext);
+
+export const AuthProvider = ({children}: {children: ReactNode}) => {
+    const [user, setUser] = useState<User | null>(null);
```
Отступы везде это 2 пробела, а не tab. Tab'ы у всех разные

reshetov commented on src/components/AuthContext.tsx:
```diff
@@ -0,0 +1,57 @@
+
+    const login = async (email: string, password: string) => {
+        const optinalUser = await findByEmail(email);
+        if (optinalUser && password === optinalUser?.password) {
```
тут safe-оператор смысла не имеет, ты шагом раньше проверяешь что optionalUser есть

reshetov commented on src/pages/Home/Home.tsx:
```diff
@@ -0,0 +1,21 @@
+  
+  return (
+      <div className='w-full text-center mt-96'>
+        <h1 className='text-4xl font-bold'>{t('welcome', { user: user?.email })}</h1>
```
Тут ? тоже лишний, у тебя этот компонент закрыть PrivateRoute, сюдя никак "не-пользователь" не зайдет

Danila-Peshev replied:
Есои уберу ts будет ругаться на 'user' is possibly 'null'

Danila-Peshev committed:
Changed all the margins to 2 spaces;
Edited the code using "Prettier";
Removed the safe operator when checking the password.
